### PR TITLE
constant interpreter: enough for basic integer operations

### PR DESCRIPTION
**Pull request by marcrasi**
Implements a constant interpreter that can deal with basic integer operations.

Summary of the features that it includes:
* builtin integer values, and builtin integer insts
* struct and tuple values, and insts that construct and extract them (necessary to use stdlib integers)
* function referencing and application (necessary to call stdlib integer functions)
* error handling data structures and logic, for telling you why your value is not evaluatable
* metatype values (not necessary for integers, but it's only a few extra lines, so I thought it would be more trouble than it's worth to put them in a separate PR)
* conditional branches (ditto)

This is a part of the https://github.com/apple/swift/pull/19579 mega-patch. See there if you're interested in what comes next.

### CODE REVIEW

File: include/swift/SILOptimizer/Utils/ConstExpr.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+//
+// This defines an interface to evaluate Swift language level constant
+// expressions.  Its model is intended to be general and reasonably powerful,
+// with the goal of standardization in a future version of Swift.
+//
+// Constant expressions are functions without side effects that take constant
+// values and return constant values.  These constants may be integer, floating
+// point, and string values, or arrays thereof (up to 1024 elements).  We allow
+// abstractions to be built out of fragile structs and tuples.
+//
```
Comment (ravikandhadai) [inline include/swift/SILOptimizer/Utils/ConstExpr.h]: @marcrasi  

> These constants may be integer, floating point, and string values, or arrays thereof (up to 1024 elements).

Should we remove this comment Line as we don't have array/string in this commit? Is the limit of 1024 elements imposed somewhere in the array extensions? If so, should we really commit to this constant?

File: lib/SIL/SILConstants.cpp [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+/// Clone this SymbolicValue into the specified allocator and return the new
+/// version.  This only works for valid constants.
+SymbolicValue
+SymbolicValue::cloneInto(llvm::BumpPtrAllocator &allocator) const {
+  auto thisRK = representationKind;
+  switch (thisRK) {
+  case RK_Unknown:
+  case RK_Metatype:
+  case RK_Function:
+  case RK_IntegerInline:
```
Comment (ravikandhadai) [inline lib/SIL/SILConstants.cpp]: @marcrasi  Seems like this function will have an assert failure inside `getIntegerValue()` if we call it with any of the first four representation kinds.  I suppose there is no reason why one should clone the first four cases as their values will probably not change at all.  In that case, can we have a default case that fails with an assert and a reasonable error message (like "cloning of this representation kind ... is not supported"? Then we can just remove the first four cases.

File: lib/SIL/SILConstants.cpp [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+
+//===----------------------------------------------------------------------===//
+// Aggregates
+//===----------------------------------------------------------------------===//
+
+/// This returns a constant Symbolic value with the specified elements in it.
+/// This assumes that the elements lifetime has been managed for this.
+SymbolicValue SymbolicValue::getAggregate(ArrayRef<SymbolicValue> elements,
+                                          llvm::BumpPtrAllocator &allocator) {
+  // Copy the integers from the APInt into the bump pointer.
```
Comment (ravikandhadai) [inline lib/SIL/SILConstants.cpp]: @marcrasi The comments are referring to ints. Probably this should be updated for arrays.

File: lib/SIL/SILConstants.cpp [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+/// to the user.  As such, walk the inlining location of the specified node to
+/// return the first location *outside* opaque libraries.
+static SILDebugLocation skipInternalLocations(SILDebugLocation loc) {
+  auto ds = loc.getScope();
+
+  if (!ds || loc.getLocation().getSourceLoc().isValid())
+    return loc;
+
+  // Zip through inlined call site information that came from the
+  // implementation guts of the tensor library.  We want to report the
```
Comment (ravikandhadai) [inline lib/SIL/SILConstants.cpp]: @marcrasi Can we rename tensor library to just library?

File: lib/SIL/SILConstants.cpp [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+    return;
+
+  std::string error;
+  switch (getUnknownReason()) {
+  case UnknownReason::Default:
+    error = "could not fold operation";
+    break;
+  case UnknownReason::TooManyInstructions:
+    // TODO: Should pop up a level of the stack trace.
+    error = "expression is too large to evaluate at compile-time";
```
Comment (ravikandhadai) [inline lib/SIL/SILConstants.cpp]: @marcrasi  Can we reword this error? Here large seems to refer to the syntactic size of the expression in the source file. It could be a small expression but may still exceed the instruction limit e.g. if it is recursive, right? Should we reword this to something more explicit like "Exceeded instruction limit: 512  when evaluating the expression at compile time"?

File: lib/SIL/SILConstants.cpp [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+  if (sourceLoc.isInvalid()) {
+    diagnose(module.getASTContext(), sourceLoc, diag::constexpr_not_evaluable);
+    return;
+  }
+  auto &SM = module.getASTContext().SourceMgr;
+  unsigned originalDiagnosticLineNumber =
+      SM.getLineNumber(fallbackLoc.getSourceLoc());
+  for (auto &sourceLoc : llvm::reverse(getUnknownCallStack())) {
+    // Skip known sources.
+    if (!sourceLoc.isValid())
```
Comment (ravikandhadai) [inline lib/SIL/SILConstants.cpp]: @marcrasi  Skip "unknown" sources?

File: lib/SILOptimizer/Mandatory/DataflowDiagnostics.cpp [hunk 7]
```diff
@@ -125,10 +165,12 @@ class EmitDFDiagnostics : public SILFunctionTransform {
       return;
 
     SILModule &M = getFunction()->getModule();
+    ConstExprEvaluator constantEvaluator(M);
```
Comment (ravikandhadai) [inline lib/SILOptimizer/Mandatory/DataflowDiagnostics.cpp]: @marcrasi  I wonder whether it would be better to create `ConstExprEvaluator` and invoke `diagnosePoundAssert` only when the experimental -enable-static-assert option is enabled.

File: lib/SILOptimizer/Utils/ConstExpr.cpp [hunk 8]
```diff
@@ -0,0 +1,730 @@
...
+// ConstExprFunctionState implementation.
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// This type represents the state of computed values within a function
+/// as evaluation happens.  A separate instance of this is made for each
+/// callee in a call chain to represent the constant values given the set of
+/// formal parameters that callee was invoked with.
+class ConstExprFunctionState {
+  /// This is the evaluator we put bump pointer allocated values into.
```
Comment (ravikandhadai) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: @marcrasi Can we make this comment a bit more clear? The evaluator holds a reference to a bump allocator and also provides utility functions like getUnknown() etc.

File: lib/SILOptimizer/Utils/ConstExpr.cpp [hunk 9]
```diff
@@ -0,0 +1,730 @@
...
+  };
+
+  // Unary operations.
+  if (inst->getNumOperands() == 1) {
+    auto operand = getConstantValue(inst->getOperand(0));
+    // TODO: Could add a "value used here" sort of diagnostic.
+    if (!operand.isConstant())
+      return operand;
+
+    // TODO: SUCheckedConversion/USCheckedConversion
```
Comment (ravikandhadai) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: @marcrasi  Is this SUCheckerConversion is what is being implemented in the following lines? Is it a TODO still?

File: lib/SILOptimizer/Utils/ConstExpr.cpp [hunk 10]
```diff
@@ -0,0 +1,730 @@
...
+// Handle calls to opaque callees, either by handling them and returning None or
+// by returning with a Unknown indicating a failure.
+llvm::Optional<SymbolicValue>
+ConstExprFunctionState::computeOpaqueCallResult(ApplyInst *apply,
+                                                SILFunction *callee) {
+  LLVM_DEBUG(llvm::dbgs() << "ConstExpr Opaque Callee: " << *callee << "\n");
+  return evaluator.getUnknown((SILInstruction *)apply, UnknownReason::Default);
+}
+
+// TODO: Refactor this to someplace common, this is defined in Devirtualize.cpp.
```
Comment (ravikandhadai) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: @marcrasi  This comment is a left-over from other patches right? It should not be removed.

File: lib/SILOptimizer/Utils/ConstExpr.cpp [hunk 11]
```diff
@@ -0,0 +1,730 @@
...
+    paramConstants.push_back(argValue);
+  }
+
+  // TODO(constexpr patch): This is currently unused, so we don't need to
+  // calculate the correct value. Eventually, include code that calculates the
+  // correct value.
+  SubstitutionMap calleeSubMap;
+
+  // Now that have successfully folded all of the parameters, we can evaluate
+  // the call.
```
Comment (ravikandhadai) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: @marcrasi  Now the "we" have

File: lib/SILOptimizer/Utils/ConstExpr.cpp [hunk 12]
```diff
@@ -0,0 +1,730 @@
...
+/// the constexpr function evaluator.  This does not handle control flow
+/// statements.  This returns None on success, and an Unknown SymbolicValue with
+/// information about an error on failure.
+llvm::Optional<SymbolicValue>
+ConstExprFunctionState::evaluateFlowSensitive(SILInstruction *inst) {
+  // These are just markers.
+  if (isa<DebugValueInst>(inst) || isa<DebugValueAddrInst>(inst) ||
+      isa<EndAccessInst>(inst) ||
+      // Constant have no important state.
+      isa<DestroyAddrInst>(inst) || isa<RetainValueInst>(inst) ||
```
Comment (ravikandhadai) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: @marcrasi  Can we just say in comments that "the interpreter doesn't model these memory management instructions and hence skips them.", instead of "Constant have no important state"

File: lib/SILOptimizer/Utils/ConstExpr.cpp [hunk 13]
```diff
@@ -0,0 +1,730 @@
...
+    }
+
+    // Otherwise, we handle terminators here.
+    if (isa<ReturnInst>(inst)) {
+      auto val = state.getConstantValue(inst->getOperand(0));
+      if (!val.isConstant())
+        return val;
+
+      // If we got a constant value, then we're good.  Set up the normal result
+      // values as well any indirect results.
```
Comment (ravikandhadai) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: as well "as"

File: lib/SILOptimizer/Utils/ConstExpr.cpp [hunk 14]
```diff
@@ -0,0 +1,730 @@
...
+/// that occur after after folding them.
+void ConstExprEvaluator::computeConstantValues(
+    ArrayRef<SILValue> values, SmallVectorImpl<SymbolicValue> &results) {
+  unsigned numInstEvaluated = 0;
+  ConstExprFunctionState state(*this, nullptr, {}, numInstEvaluated);
+  for (auto v : values) {
+    auto symVal = state.getConstantValue(v);
+    results.push_back(symVal);
+
+    // Reset the execution limit back to zero for each subsexpression we look
```
Comment (ravikandhadai) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: subsexpression -> subexpression

File: lib/SILOptimizer/Utils/ConstExpr.cpp [hunk 15]
```diff
@@ -0,0 +1,730 @@
...
+}
+
+/// Analyze the specified values to determine if they are constant values.  This
+/// is done in code that is not necessarily itself a constexpr function.  The
+/// results are added to the results list which is a parallel structure to the
+/// input values.
+///
+/// TODO: Return information about which callees were found to be
+/// constexprs, which would allow the caller to delete dead calls to them
+/// that occur after after folding them.
```
Comment (ravikandhadai) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: after after -> after

File: test/SILOptimizer/pound_assert.swift [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+//===----------------------------------------------------------------------===//
+
+func test_topLevelEvaluation(topLevelArgument: Int) {
+  let topLevelConst = 1
+  #assert(topLevelConst == 1)
+
+  // TODO: I think that this should get rejected. Otherwise, constness of vars
+  // depends on what you do to them, which is confusing. Perhaps some other
+  // const folder is getting to this before ConstExpr.cpp does?
+  var topLevelVar = 1 // expected-warning {{never mutated}}
```
Comment (ravikandhadai) [inline test/SILOptimizer/pound_assert.swift]: @marcrasi The answer to the question in the comments is Yes. The `DiagnosticConstantPropagation` phase runs before this phase and it folds this constant. This I think should be okay, it only makes #assert's a little bit more powerful ;-)

File: include/swift/SIL/SILConstants.h [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+    RK_Function,
+
+    /// This value is represented with a bump-pointer allocated APInt.
+    RK_Integer,
+
+    /// This value is represented with an inline integer representation.
+    RK_IntegerInline,
+
+    /// This value is an array, struct, or tuple of constants.  This is
+    /// tracked by the "aggregate" member of the value union.
```
Comment (ravikandhadai) [inline include/swift/SIL/SILConstants.h]: @marcrasi  "This value is an array, ..." -> For arrays, we will have a different representation kind, right? Guess array should be removed from this comment.

File: lib/SILOptimizer/Utils/ConstExpr.cpp (hunk 10 above)
Comment (marcrasi) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: yeah, it shouldn't be in this patch. will remove.

**Comment by marcrasi** [comment]
I'm partway through rebasing this onto latest changes and addressing your comments, so I replied to some and marked some as resolved. I haven't uploaded my changes yet so no need to take a look or respond yet. I'm going to stop now and do a bit more tomorrow.

File: lib/SILOptimizer/Mandatory/DataflowDiagnostics.cpp (hunk 7 above)
Comment (marcrasi) [reply lib/SILOptimizer/Mandatory/DataflowDiagnostics.cpp]: yes, that sounds good. will do.

File: test/SILOptimizer/pound_assert.swift [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+//===----------------------------------------------------------------------===//
+// Integers
+//===----------------------------------------------------------------------===//
+
+func test_trapsAndOverflows() {
+  // The error message below is generated by the traditional constant folder.
+  // expected-error @+1 {{arithmetic operation '124 + 92' (on type 'Int8') results in an overflow}}
+  #assert((124 as Int8) + 92 < 42)
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply test/SILOptimizer/pound_assert.swift]: It seems that the other constant folder is replacing the condition in this case, but leaving the condition alone in the next case. I have added some comments explaining that this is happening, but I haven't investigated why the other constant folder is behaving like that.

File: lib/SILOptimizer/Utils/ConstExpr.cpp [hunk 19]
```diff
@@ -0,0 +1,730 @@
...
+    // TODO: SUCheckedConversion/USCheckedConversion
+
+    // Implement support for s_to_s_checked_trunc_Int2048_Int64 and other
+    // checking integer truncates.  These produce a tuple of the result value
+    // and an overflow bit.
+    //
+    // TODO: We can/should diagnose statically detectable integer overflow
+    // errors and subsume the ConstantFolding.cpp mandatory SIL pass.
+    auto IntCheckedTruncFn = [&](bool srcSigned,
+                                 bool dstSigned) -> SymbolicValue {
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: Yes, that sounds like an improvement. fyi, dstSigned already was used on the line that calculates `reextended`.

File: lib/SILOptimizer/Utils/ConstExpr.cpp (hunk 9 above)
Comment (marcrasi) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: It's still a TODO. The following lines implement "xxxCheckedTrunc", not "xxxCheckedConversion".

"xxxCheckedConversion" seems to be a builtin that exists but that the stdlib never calls.

**Comment by marcrasi** [comment]
Okay, I have rebased this and addressed this round of comments!

For the rebase, I had to change SILConstants.h to accept an ASTContext instead of a BumpPtrAllocator because ASTContext's getAllocator() became private.

File: test/SILOptimizer/pound_assert.swift (hunk 18 above)
[the remark this answers is not in the document]
Comment (ravikandhadai) [reply test/SILOptimizer/pound_assert.swift]: @marcrasi  I see. Thanks for the clarification. Probably the diagnostic constant propagation, in this case, replaces the addition with a truncated value (besides recording the diagnostics). So the interpreter never sees the addition. Btw, a minor comment is that I think it would be better in these comments to use the term "interpreter"  instead of constant folder for #assert. It may create less confusion if we use the terminology "folder" for something that actually changes the code, whereas the interpreter doesn't.  We may want to eventually rewrite the constant folder using the interpreter, but still I think the folding will be done by the constant folder phase.

**Comment by ravikandhadai** [comment]
@marcrasi  Thanks for making these changes. It looks good to me.

**Review by ravikandhadai — APPROVED** [review]

File: test/SILOptimizer/pound_assert.swift (hunk 18 above)
[the remark this answers is not in the document]
Comment (marcrasi) [reply test/SILOptimizer/pound_assert.swift]: That sounds like a good rewording. I have updated the comment.

**Comment by marcrasi** [comment]
I rebased this one to master now that the previous part is merged. Is this ready to test and merge?

**Comment by ravikandhadai** [comment]
@marcrasi  Thanks Marc. It would be good to have @devincoughlin  review. Let us wait a bit for that. Let me run the tests in the meantime.

**Comment by ravikandhadai** [comment]
@swift-ci Please test

**Comment by ravikandhadai** [comment]
@swift-ci Please test macOS platform

**Comment by ravikandhadai** [comment]
@swift-ci Please benchmark

**Review by devincoughlin — COMMENTED** [review]
This looks good to me. Some minor requests inline and a suggestion to add a SIL-level test.

File: lib/SIL/SILConstants.cpp [hunk 20]
```diff
@@ -0,0 +1,405 @@
...
+  unsigned originalDiagnosticLineNumber =
+      SM.getLineNumber(fallbackLoc.getSourceLoc());
+  for (auto &sourceLoc : llvm::reverse(getUnknownCallStack())) {
+    // Skip unknown sources.
+    if (!sourceLoc.isValid())
+      continue;
+    // Also skip notes that point to the same line as the original error, for
+    // example in:
+    //   #assert(foo(bar()))
+    // it is not useful to get three diagnostics referring to the same line.
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply lib/SIL/SILConstants.cpp]: I can't think of a situation where this logic does something useful, so I'll just remove it.

The example in the comments seems to suggest that `#assert(foo(bar()))` will emit two notes when `bar()` is not const-evaluable: one for `bar` and one for `foo`. But that is not actually what happens -- it seems that the unknown that comes out of `bar()` is the one that emits the notes, so only a note for `bar()` gets emitted. And even if it did emit two notes, I agree that that would be reasonable.

File: lib/SIL/SILConstants.cpp [hunk 21]
```diff
@@ -0,0 +1,405 @@
...
+  auto &module = badInst->getModule();
+  diagnose(module.getASTContext(), loc.getSourceLoc(),
+           diag::constexpr_unknown_reason, error)
+      .highlight(loc.getSourceRange());
+  return true;
+}
+
+/// Given that this is an 'Unknown' value, emit diagnostic notes providing
+/// context about what the problem is.
+void SymbolicValue::emitUnknownDiagnosticNotes(SILLocation fallbackLoc) {
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply lib/SIL/SILConstants.cpp]: I don't want to move this logic into a pass because there may be multiple different passes that want to use it to emit diagnostics about unknown values. (e.g. some of the tensorflow passes need to const-eval certain things and emit diagnostics when the const-eval fails).

Having it in SymbolicValue seems like the easiest way for passes who use it to call it, and I don't see much harm in leaving it here. Alternatively, I could turn it into a standalone utility function. Maybe `emitUnknownDiagnosticNotes(SILConstant unknown, SILLocation fallbackLoc)` in `SILOptimizer/Utils/ConstantDiagnostics`? What do you prefer?

File: lib/SIL/SILConstants.cpp [hunk 22]
```diff
@@ -0,0 +1,395 @@
...
+  auto badInst = dyn_cast<SILInstruction>(getUnknownNode());
+  if (!badInst)
+    return;
+
+  std::string error;
+  switch (getUnknownReason()) {
+  case UnknownReason::Default:
+    error = "could not fold operation";
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply lib/SIL/SILConstants.cpp]: will do

File: lib/SILOptimizer/Utils/ConstExpr.cpp [hunk 23]
```diff
@@ -0,0 +1,725 @@
...
+
+public:
+  ConstExprFunctionState(ConstExprEvaluator &evaluator, SILFunction *fn,
+                         SubstitutionMap substitutionMap,
+                         unsigned &numInstEvaluated)
+      : evaluator(evaluator), fn(fn), substitutionMap(substitutionMap),
+        numInstEvaluated(numInstEvaluated) {}
+
+  void setValue(SILValue value, SymbolicValue symVal) {
+    calculatedValues.insert({value, symVal});
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: will do. actually, `Address` kinds don't exist until the next PR (https://github.com/apple/swift/pull/19657/files) so I'll just add a commented-out assertion with a TODO to uncomment it.

File: include/swift/SIL/SILConstants.h [hunk 24]
```diff
@@ -0,0 +1,250 @@
...
+
+  /// Clone this SymbolicValue into the specified ASTContext and return the new
+  /// version.  This only works for valid constants.
+  SymbolicValue cloneInto(ASTContext &astContext) const;
+
+  void print(llvm::raw_ostream &os, unsigned indent = 0) const;
+  void dump() const;
+};
+
+static_assert(sizeof(SymbolicValue) == 2 * sizeof(void *),
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply include/swift/SIL/SILConstants.h]: will do

File: test/SILOptimizer/pound_assert.swift [hunk 25]
```diff
@@ -0,0 +1,155 @@
+// RUN: %target-swift-frontend -enable-experimental-static-assert -emit-sil %s -verify -Xllvm -debug -Xllvm -debug-only -Xllvm ConstExpr
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply test/SILOptimizer/pound_assert.swift]: will do

File: lib/SIL/SILConstants.cpp [hunk 26]
```diff
@@ -0,0 +1,405 @@
...
+APInt SymbolicValue::getIntegerValue() const {
+  assert(getKind() == Integer);
+  if (representationKind == RK_IntegerInline) {
+    auto numBits = aux.integer_bitwidth;
+    return APInt(numBits, value.integerInline);
+  }
+
+  assert(representationKind == RK_Integer);
+  auto numBits = aux.integer_bitwidth;
+  auto numWords = (numBits + 63) / 64;
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply lib/SIL/SILConstants.cpp]: will do

File: include/swift/SIL/SILConstants.h [hunk 27]
```diff
@@ -0,0 +1,246 @@
...
+    /// When this SymbolicValue is of "Aggregate" kind, this pointer stores
+    /// information about the array elements and count.
+    const SymbolicValue *aggregate;
+  } value;
+
+  RepresentationKind representationKind : 8;
+
+  union {
+    /// This is the reason code for RK_Unknown values.
+    UnknownReason unknown_reason : 32;
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply include/swift/SIL/SILConstants.h]: will do

File: include/swift/SIL/SILConstants.h [hunk 28]
```diff
@@ -0,0 +1,246 @@
...
+
+    /// This is the number of bits in an RK_Integer or RK_IntegerInline
+    /// representation, which makes the number of entries in the list derivable.
+    unsigned integer_bitwidth;
+
+    /// This is the number of elements for an RK_Aggregate representation.
+    unsigned aggregate_numElements;
+  } aux;
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply include/swift/SIL/SILConstants.h]: will do

File: lib/SIL/SILConstants.cpp [hunk 29]
```diff
@@ -0,0 +1,395 @@
...
+  friend class llvm::TrailingObjects<UnknownSymbolicValue, SourceLoc>;
+
+  /// The value that was unfoldable.
+  SILNode *node;
+
+  /// A more explanatory reason for the value being unknown.
+  UnknownReason reason;
+
+  /// The number of elements in the call stack.
+  unsigned call_stack_size;
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply lib/SIL/SILConstants.cpp]: will change

**Comment by ravikandhadai** [comment]
@swift-ci please smoke test

File: lib/SILOptimizer/Utils/ConstExpr.cpp [hunk 30]
```diff
@@ -0,0 +1,725 @@
...
+                                                SILFunction *callee) {
+  LLVM_DEBUG(llvm::dbgs() << "ConstExpr Opaque Callee: " << *callee << "\n");
+  return evaluator.getUnknown((SILInstruction *)apply, UnknownReason::Default);
+}
+
+/// Given a call to a function, determine whether it is a call to a constexpr
+/// function.  If so, collect its arguments as constants, fold it and return
+/// None.  If not, mark the results as Unknown, and return an Unknown with
+/// information about the error.
+llvm::Optional<SymbolicValue>
```
[the remark this answers is not in the document]
Comment (devincoughlin) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: Yes, that sounds good.

**Comment by devincoughlin** [comment]
Looks good to me. Thanks!

**Review by devincoughlin — APPROVED** [review]

**Comment by ravikandhadai** [comment]
@swift-ci Please test

**Comment by ravikandhadai** [comment]
@swift-ci please test and merge
